docs/reference/search/search-your-data/run-knn-search.asciidoc

```diff
@@ -0,0 +1,248 @@
+large datasets. The vector function has to scan each matched document linearly,
+resulting in slow search speeds. However, you can improve latency by using a
+query to limit the number of matched documents passed to the function.
+
+[discrete]
+[[approximate-knn]]
+=== Approximate kNN
+
+To run an approximate kNN search, use the kNN search API to search a
+`dense_vector` field with indexing enabled.
+
+. Explicitly map one or more `dense_vector` fields. Approximate kNN search
+requires the following mapping options:
++
+--
+* An `index` value of `true`.
+
+* A `similarity` value. This value determines the similarity metric used to
+score documents based on similarity between the query and document vector. For a
+list of available metrics, see the <<dense-vector-similarity,`similarity`>>
+parameter documentation.
+
+include::{es-repo-dir}/mapping/types/dense-vector.asciidoc[tag=dense-vector-indexing-speed]
+
+[source,console]
+----
+PUT my-approx-knn-index
+{
+  "mappings": {
+    "properties": {
+      "my-image-vector": {
+        "type": "dense_vector",
+        "dims": 5,
+        "index": true,
+        "similarity": "l2_norm"
+      },
+      "my-tag": {
+        "type": "keyword"
+      }
+    }
+  }
+}
+----
+--
+
+. Index your data.
++
+[source,console]
+----
+POST my-approx-knn-index/_bulk?refresh=true
+{ "index": { "_id": "1" } }
+{ "my-image-vector": [230.0, 300.33, -34.8988, 15.555, -200.0], "my-tag": "cow.jpg" }
+{ "index": { "_id": "2" } }
+{ "my-image-vector": [-0.5, 100.0, -13.0, 14.8, -156.0], "my-tag": "moose.jpg" }
+{ "index": { "_id": "3" } }
+{ "my-image-vector": [0.5, 111.3, -13.0, 14.8, -156.0], "my-tag": "rabbit.jpg" }
+...
+----
+//TEST[continued]
+//TEST[s/\.\.\.//]
+
+. Run the search using the <<knn-search,kNN search API>>.
++
+[source,console]
+----
+GET my-approx-knn-index/_knn_search
+{
+  "knn": {
+    "field": "my-image-vector",
+    "query_vector": [-0.5, 90.0, -10, 14.8, -156.0],
+    "k": 10,
+    "num_candidates": 100
+  },
+  "fields": [
+    "my-image-vector",
+    "my-tag"
+  ]
+}
+----
+//TEST[continued]
+// TEST[s/"k": 10/"k": 3/]
+// TEST[s/"num_candidates": 100/"num_candidates": 3/]
+
+[discrete]
+[[tune-approximate-knn-for-speed-accuracy]]
+==== Tune approximate kNN for speed or accuracy
+
+For faster searches, the kNN search API collects `num_candidates` results from
+each shard, then merges them to find the top `k` results. This works well in
+most cases, but results are less accurate if a shard contains a disproportionate
+number of nearest neighbors.
+
+You can increase `num_candidates` for more accurate results at slower search
+speeds. Similarly, you can decrease `num_candidates` for faster searches with
+potentially less accurate results.
+
+[discrete]
+[[approximate-knn-limitations]]
+==== Limitations for approximate kNN search
+
+* You can't use Query DSL to filter the documents searched using approximate kNN
+or the results of an approximate kNN search. If you need to filter documents,
```

Comment:
> If you need to filter documents,

May be modify to "if you need to filter documents with restrictive filters"

---
As if a filter is not restrictive, it is still worth to do approximate kNN, and do filtering after knn search on the user application side.

Reply:
Thanks for pointing this out. I reworded to:

> If you need to restrictively filter the documents on which a kNN search runs, consider using exact kNN instead.